### DISCUSSION TITLE
A new way to wrap eglot's CAPF (with some help)

### DISCUSSION THREAD

**Discussion by jdtsmith** [minad/cape#106] [Ideas]
A discussion on emacs-devel gave me an idea for a new and better way to interact with eglot's CAPF.  It's a remix of various approaches we've tried.  The solution:

1. A custom orderless style dispatcher which passes all matches _unfiltered_ from the first orderless term (I already do this), and 
2. A cape buster wrapper which causes the CAPF to "continuously update" completions as the first term (only) gets edited, ignoring all other terms.  

This combines the best of both worlds: continuous update of the starting term, and orderless based filtering "after the space" (entered in corfu as `M-SPC`).  It's easy to know which "side" of the table-dumping just-ask-the-LSP-server fence you are on: are your edits before or after the space?  

So something like this _should_ do the trick:

```elisp
(cl-pushnew
 (lambda (pattern index &rest r) (when (eq index 0) "")) ; "" = "all candidates"
     orderless-style-dispatchers)
(cl-nsubst
 (cape-capf-buster
  #'eglot-completion-at-point
  (cl-labels ((not-space (x) (/= x ?\s)))
    (lambda (old new)	   ; valid if everything up to space unchanged
      (equal (seq-take-while #'not-space old)
	     (seq-take-while #'not-space new)))))
 'eglot-completion-at-point
 completion-at-point-functions)
```

But alas, no luck.  Then, in `cape-wrap-buster` I see:

```elisp
              (unless (or (string-match-p "\\s-" new-input) ;; Support Orderless
                          (funcall valid input new-input))
```

Ahah, this is actively thwarting multi-orderless-by-term busting, which is exactly what I needed.  So I removed that.  But _still_ no dice.  Then I see:

```elisp
(guard (and (= beg new-beg) (= end new-end)))
```

At first, I think: _this must be a bug_.  If you are editing the search term (adding/deleting), you'd very much expect the LSP server to return a _new_ `beg` and `end`.  So I think this should be simply `(guard (= beg new-beg))`.    But _then_ I see you're using markers for `beg` and `end` as closure variables.  Clever!  But actually, _too_ clever, when orderless terms are used.  That's because `end` moves ahead with `M-SPC`, but the eglot CAPF does not consider anything past the space as relevant for completion, so if you come bak and edit the first term, `beg` and `end` are mismatched, and from then on, no matter what the `valid` function has to say, the capf is no longer busted.

One solution would be to ensure `M-SPC` leaves the `end` marker behind (I'm not sure why it wouldn't atm).  But perhaps better is just `(guard (= beg new-beg))`, which is IMO safe enough.

With those in place, the experience is _quite_ excellent.  Within the first term, eglot behaves "just like company".  After the first term, it behaves "just like orderless".  `M-SPC` is the clear delineator between these two "modes of operation", and you can happily edit on "either side" and it does the right thing.

Would you consider an optional argument to `cape-capf-buster` to disable the orderless specific validity assertion?  And what do you think about guarding only the CAPF's `beg` against motion?

**Comment by minad** [comment]
I am sorry. I cannot understand what you wrote before having a coffee. The idea is that the Capf buster only updates the completion table as long as no Orderless separator has been entered. While this is the case beg and end validity is guaranteed. What is different in your proposal? (Btw please update you version of Cape and Corfu, since the code you cite is outdated).

**Reply by minad** [reply]
I've reread your comment again and I still don't have a clue. Could it be that you misunderstood the code of `cape-capf-super` (the condition where we check if the table should be refreshed) or that there has been a bug in the old version? Please check again the newest version.

**Comment by minad** [comment]
I looked at the thread you mentioned. It looks as if João appears to be more positive to the caching behavior used by Corfu and Icomplete:

https://lists.gnu.org/archive/html/emacs-devel/2024-01/msg01208.html

> I've turned the idea down in the past, but if you think it can
contribute to your eglot-c-a-p overhaul AND it doesn't break Company and
other things, I'm willing to reconsider my opinion.  Especially now that
icomplete-in-region is a thing and _I think_ it also assumes completion
tables behaves like JD/Corfu likes them to.  JD can lead the way here,
if he's available.

If this happens we can hopefully ditch `cape-capf-buster` for good. Note that Corfu already ensures that the completion predicate is not rechecked if Orderless separation is used. 

So please lets focus on improving the Eglot Capf instead of trying to change the semantics of `cape-capf-buster`.

**Reply by jdtsmith** [reply]
I hadn't actually seen that bit, very interesting.  There may of course be other CAPFs that need busting, but very good that the arguments eventually landed.

Apologies that what I wrote wasn't clear (written too late and too quickly, in the midst of other things). 

What I'm after here is a CAPF behavior that works proactively _with_ orderless, by updating candidates from the server _only_ when the first orderless term changes, and not otherwise.  This builds on the idea of how `cape-wrap-buster` can take a `valid` argument of say `#'equal` vs. `#'string-prefix-p`.  The former doesn't work with orderless (since orderless search terms are meaningless to the LSP server).  The latter depends on the random point where corfu-auto got called, so it's not great either.  

So I devised a "smarter" `valid` predicate that simply tests for string equality _only up to the first space character_, i.e. only of the first orderless term.  But `cape-wrap-buster` currently defeats that, because it (as you mention) explicitly checks for a space in the candidate, and asserts validity if it finds one there: no more busting (= LSP server re-querying).  And then it defeats it _again_ by insisting the positional bounds returned by the CAPF are always unchanged (modulo marker movements), which condition does not hold when editing the first match of a `full __orderless space-separated ![a-z]match`, because the CAPF correctly ignores things after the 1st space.  

This is of course too much to ask EGLOT (or any capf) to do on its own. But, when it is working (with my suggested edits above), the way it works in one continuous corfu-auto session is quite magical (to me):

```python
big.[_]     # large list of big module's completions ([_]=corfu-auto at work)
big.at      # pretty big list of LSP-provided completions, whatever it thinks (attach, abstract, last, all manner of matches...)
big.at !__  # via M-SPC, remove all __ methods and variables. LSP server left un-queried
big.apt !__ # re-queries the LSP server, and also orderless filters the results; still in the same completion session
```

**Reply by jdtsmith** [reply]
This all makes me wonder if there shouldn't be a bit more intelligence possible in the chain.  Right now we have two situations:

1. Some UI's (like company) take control and aggressively dump completion tables, treating the CAPFs a bit like mis-behaving children, trying to get them to do the right thing and keep completion results fresh.
2. Others, like corfu, assume the best of CAPFs, and leave it up to their completion tables, trusting them to do the right thing on their own.

That leaves the completion _style_ out in the cold, without any say in the matter.  I can imagine a completion style like orderless offering an opinion to the CAPF: "please keep your candidates fresh, but don't worry about anything after the space, that's my domain."

**Reply by minad** [reply]
I still don't understand. Could you please prepare a patch which shows the change you have in mind? And also list two test cases which demonstrate the difference in behavior.

I claim that the current behavior is reasonable. While we don't have a separator we can refresh the table, since the backend doesn't have to know about Orderless. Note that we only check the valid function if no separator is present. At the same time we can require that the begin and end bounds stay intact. This is consistent with Corfu which also checks the completion predicate if no separator has been inserted. Note that the end check uses a marker, which moves on input, so I don't see how this defeats something.

**Reply by minad** [reply]
> What I'm after here is a CAPF behavior that works proactively with orderless, by updating candidates from the server only when the first orderless term changes, and not otherwise. 

What you write here - this is the current behavior we already have. At least that's my assumption and this is why I am confused.

**Reply by jdtsmith** [reply]
> Note that we only check the valid function if no separator is present.

Yes, and that _is the problem_.  I want the `valid` function to be in charge of that decision.  Let me see if I can spin up a `wrap-orderless-eglot` to show the difference (can't get to until late tonight).

**Comment by jdtsmith** [comment]
> > What I'm after here is a CAPF behavior that works proactively with orderless, by updating candidates from the server only when the first orderless term changes, and not otherwise.
> 
> What you write here - this is the current behavior we already have. At least that's my assumption and this is why I am confused.

Well no, the current behavior is to stop checking for validity as soon as it's clear to the wrapper that it's an orderless search.  I'm after something different: _keep_ checking for validity, and declare the current table invalid _ONLY_  when the **VERY FIRST** orderless term (`[this] term right here`) changes.  In the current version, once an orderless search is detected, the table will never be invalidated.  In my tweak, if I edit it to `[that] term right here`, the table will be invalidated, LSP will be re-queried (with `that`), and orderless will work as intended, on top.

**Reply by minad** [reply]
Okay, I understand. This cannot work since the lsp client will always see the full input `[that] term right here`.

This is part of the often cited impedance mismatch of lsp and emacs completion. The buffer input is what counts and not the prefix/regexps passed down by the completion style to the table.

**Comment by jdtsmith** [comment]
> Okay, I understand. This cannot work since the lsp client will always see the full input `[that] term right here`.

No, it doesn't.  Because LSP only examines the buffer state (it cares naught for what completion-in-region thinks is the "current text"), and completion doesn't span spaces.

**Reply by minad** [reply]
Sorry, no. I didn't express myself clearly enough. Lsp will see the full buffer text including the full input. This means it will get confused by the additional orderless tokens. Your proposal will not work.

**Reply by minad** [reply]
Thinking more about this, it may work if the server ignores all the additional tokens which it may do. We could move the entire responsibility of refreshing to the validator function. But then we still have the problem with the begin/end check.

Note that cape-company-to-capf and cape--dynamic-table would also be affected by such changed semantics. These functions use the prefix to compute the candidates and the rest of the input only for filtering. They could as well refresh if only the prefix changes, instead of bailing out if a separator space is detected.

**Reply by minad** [reply]
Thinking even more about this, it seems to be that the behavior will be point-dependent. The point is marked by |, and boundaries by < and >.

Situation 1 (all good):

1. `object.<prefix other words|>`
2. `object.<changed| other words>` -> table will be rerequested and fully filtered by the whole input.

Situation 2 (will break):

1. `object.<prefix other words|>`
2. `object.<changed o|ther words>` -> new table is requested, but the returned results will refer to `other` and not to `changed`.

**Reply by minad** [reply]
I tried to implement the modified validation here: https://github.com/minad/cape/compare/relax-refresh-condition

Can you please give it a test and maybe check the code if it looks reasonable to you?

**Reply by jdtsmith** [reply]
Looks perfect at a glance.  Can give a try tonight.

**Reply by jdtsmith** [reply]
OK I tested the new branch for a short while.  At first blush, it seems to work well, similar to what I had before.  The problem I eventually encountered is that my server cares where point is very much, and hence _where_ you make your edits.  I.e. `<pref|ix>` and `<prefix|>` generate different tables (it ignores text after point).  This is the same thing you'd get if you `M-Tab` in those locations, so it shouldn't have been a surprise.  I need to keep playing with it a bit.  If you delete over your prefix first and then start typing, it works great. 

**Update** This actually has nothing to do with this issue or the fix you put in.  It applies to normal `#'equal`-based capf-busting too.  If you go back and edit your search term in the middle, the server will ignore the stuff after point and give you completions for stuff up to point only.  I don't know if this is an eglot issue or a basic feature of the LSP protocol, or particular to my lsp server.   I do know you can send a "range" position instead of just line/col.  So I think this fix is working about as well as it could given that limitation.  In any case, much more useful than the old "just ignore any changes when there's a space" approach.

**Reply by minad** [reply]
Thanks for checking.

> So I think this fix is working about as well as it could given that limitation. In any case, much more useful than the old "just ignore any changes when there's a space" approach.

I am not sure I agree with this. The behavior is less consistent with this change. The part of the prefix after point is ignored while the part after space is not ignored because of completion style filtering. This whole thing seems like a marginal improvement or not even. A lot of discussion for nothing. The existing behavior is simple - delete up to the prefix you want to complete, take it from there and new candidates are generated reliably. The new behavior still requires you to do deletion in the prefix, or care very much about your point placement.

**Reply by minad** [reply]
Note that I also tested cape-dict which does not have such a point position limitation, but which only generates a small number of candidates. If you enter Orderless search terms and then edit the prefix you may not get any candidates at all. Not even there I am convinced that the update behavior is better. Deleting up to the prefix and starting anew is more reliable.

**Reply by jdtsmith** [reply]
> This whole thing seems like a marginal improvement or not even. A lot of discussion for nothing. The existing behavior is simple - delete up to the prefix you want to complete, take it from there and new candidates are generated reliably. The new behavior still requires you to do deletion in the prefix, or care very much about your point placement.

How you delete the prefix is not really related to the main improvement, since it impacts any completion style.  No matter what, `prefix` => `pre[d]ix` will give completions without an `x`.  

But it's definitely not _nothing_.  The main improvement here is allowing prefix alterations _at all_ when you have orderless filtering there.  The old buster simply did not allow this at all.   Not a life-changing improvement, but definitely helpful at times.

**Comment by jdtsmith** [comment]
> Situation 2 (will break):
> 
> 1. `object.<prefix other words|>`
> 2. `object.<changed o|ther words>` -> new table is requested, but the returned results will refer to `other` and not to `changed`.



But this is the key insight: situation 2 will _not_ break (at least with my server I've already verified this).  Because the new table will _not_ be requested, since the edit occurs among the `other words`, and the `valid` function is only comparing the first term.

**Reply by jdtsmith** [reply]
I mean you could do some kind of insertion _not_ at point, but that's going to mess with Eglot no matter what.  Edits to `prefix` will have point in `prefix`; edits in `other words` will have point in `other words`.

**Reply by minad** [reply]
Sorry, I do absolutely not believe this key insight. There is again a misunderstanding. 

What I mean is that point is inside the word `other`. At this point, the server will surely not return candidates which match the prefix `changed`. Note that this is timing dependent. If you change `prefix` quickly to `changed` *and then move point quickly to the right*, before Corfu refreshes the completion table, the candidates won't be correct. We would get candidates based on `other` and not based on `changed` as we actually want. Therefore we have to guard against this, and check if point lies inside the prefix, see:

https://github.com/minad/cape/commit/c8d5472739304fd045fee3959ccd68d7366a75a5#diff-e49573be46636710c6be9a8fe36c6507dd4c729235d5d91c0c2a994ecc948f06R1013-R1015

I am on the fence regarding this change. The current behavior is overall simpler to reason about. Otherwise we would not repeatedly misunderstand each other or make mistakes in our argumentation.

However the change in https://github.com/minad/cape/compare/relax-refresh-condition itself is simple enough so I am willing to consider it.

**Reply by minad** [reply]
> I mean you could do some kind of insertion not at point, but that's going to mess with Eglot no matter what. Edits to prefix will have point in prefix; edits in other words will have point in other words.

It is not only about edits at a different location. The problem also occurs if the point is moved quickly.

Also the relaxed check of new-beg and new-end basically calls for problems:

https://github.com/minad/cape/commit/c8d5472739304fd045fee3959ccd68d7366a75a5#diff-e49573be46636710c6be9a8fe36c6507dd4c729235d5d91c0c2a994ecc948f06R1023-R1024

**Reply by jdtsmith** [reply]
> Sorry, I do absolutely not believe this key insight. There is again a misunderstanding.
> 
> What I mean is that point is inside the word `other`. At this point, the server will surely not return candidates which match the prefix `changed`. Note that this is timing dependent. If you change `prefix` quickly to `changed` _and then move point quickly to the right_, before Corfu refreshes the completion table, the candidates won't be updated. 

Ah, right.  I guess corfu is the one kicking this off in its timer.  This should be OK though, since it's the 1st-term which is always being compared; as soon as that is noticed, the table will be refreshed (and orderless will filter the results using the `other words`).  If timing issues do appear, it will be no worse than before, where the `valid` function just stopped having any effect as soon as there was a space involved, right?  

> Or if they would be updated, we would get candidates based on `other` and not based on `changed` as we actually want. Therefore we have to guard against this, and check if point lies inside the prefix, see:
> 
> [c8d5472#diff-e49573be46636710c6be9a8fe36c6507dd4c729235d5d91c0c2a994ecc948f06R1013-R1015](https://github.com/minad/cape/commit/c8d5472739304fd045fee3959ccd68d7366a75a5#diff-e49573be46636710c6be9a8fe36c6507dd4c729235d5d91c0c2a994ecc948f06R1013-R1015)
> 
> I am on the fence regarding this change. The current behavior is overall simpler to reason about. Otherwise we would not repeatedly misunderstand each other or make mistakes in our argumentation.

I suppose so.  It's very specific to 1) a table which needs to be invalidated/refreshed from the outside, 2) using orderless in buffer.  But I don't think it's too complex, probably just me not taking the time to explain it carefully enough on the 1st attempt. 

> However the change in https://github.com/minad/cape/compare/relax-refresh-condition itself is simple enough so I am willing to consider it.

:) Looks simple to me. And likely to work well.

**Reply by minad** [reply]
To be fair the current behavior also has a timing dependence. If I change the prefix and then enter a space quickly enough, the completion candidates will also not get updated.

**Reply by minad** [reply]
> Ah, right. I guess corfu is the one kicking this off in its timer. This should be OK though, since it's the 1st-term which is always being compared; as soon as that is noticed, the table will be refreshed (and orderless will filter the results using the other words). 

Not quite. It doesn't suffice to notice a prefix change. Point also has to move back into the prefix, since otherwise we would get wrong candidats. See the guard: https://github.com/minad/cape/commit/c8d5472739304fd045fee3959ccd68d7366a75a5#diff-e49573be46636710c6be9a8fe36c6507dd4c729235d5d91c0c2a994ecc948f06R1013-R1015

> If timing issues do appear, it will be no worse than before, where the valid function just stopped having any effect as soon as there was a space involved, right?

Yes, things won't be worse I suppose.

**Reply by minad** [reply]
The question is if we could move point back to the end of the prefix and then restore it after refreshing the table. I don't know enough about the Eglot Capf to tell if this would work.

EDIT: See the commit  https://github.com/minad/cape/commit/6482b772c9ac1b888a70114abc7de2a79138cd24. I believe this eliminates the timing dependence.

**Reply by jdtsmith** [reply]
> EDIT: See the commit  https://github.com/minad/cape/commit/6482b772c9ac1b888a70114abc7de2a79138cd24. I believe this eliminates the timing dependence.

Excellent, looks like a good approach.  

Also, am I understanding correctly that even during active corfu completion it's the same timer running _corfu-auto_ which kicks off the table refresh?  I guess it has to be, since who else calls the CAPF table.  Somehow I had actually assumed that (like eglot's update to the lsp server itself) this was happening in an after-change-function.  Sort of dumb of me in hindsight.

**Reply by minad** [reply]
> Also, am I understanding correctly that even during active corfu completion it's the same timer running corfu-auto which kicks off the table refresh? I guess it has to be, since who else calls the CAPF table. Somehow I had actually assumed that (like eglot's update to the lsp server itself) this was happening in an after-change-function. Sort of dumb of me in hindsight.

No, it is a little more subtle than that. The update happens inside a post-command-hook, but the update can get interrupted if you type fast enough. There is no timer, but still a time dependence and a dependence on your typing speed.

**Reply by jdtsmith** [reply]
OK, so I was being less dumb than expected.  How encouraging.
